internal/grpc/commander/connection.go

```diff
@@ -53,17 +84,138 @@ func newConnection(
+		configUpdatedCh:       make(chan struct{}, configUpdatedChSize),
+		configApplyResponseCh: make(chan configApplyResponse),
+		id:                    id,
+	}
+}
+
+func (c *connection) ID() string {
+	return c.id
+}
+
+func (c *connection) State() State {
+	return c.state
+}
+
+func (c *connection) Update() {
+	select {
+	case c.configUpdatedCh <- struct{}{}:
+		c.logger.Info("Queued config update")
+	default:
+	}
+}
+
+func createDownloadCommand(msgID, systemID, nginxID string) *proto.Command {
+	return &proto.Command{
+		Meta: &proto.Metadata{
+			MessageId: msgID,
+		},
+		Type: proto.Command_DOWNLOAD,
+		Data: &proto.Command_NginxConfig{
+			NginxConfig: &proto.NginxConfig{
+				Action: proto.NginxConfigAction_APPLY,
+				ConfigData: &proto.ConfigDescriptor{
+					SystemId: systemID,
+					NginxId:  nginxID,
+				},
+			},
+		},
+	}
+}
+
+func (c *connection) sendConfig(request *proto.DownloadRequest, downloadServer proto.Commander_DownloadServer) error {
+	config := c.pendingConfig
+
+	if config.ID != request.GetMeta().GetMessageId() {
+		err := fmt.Errorf(
+			"pending config ID %q does not match request %q",
+			config.ID,
+			request.GetMeta().GetMessageId(),
+		)
+		c.logger.Error(err, "failed to send config")
+		return err
+	}
+
+	cfg := &proto.NginxConfig{
+		Action: proto.NginxConfigAction_APPLY,
+		ConfigData: &proto.ConfigDescriptor{
+			SystemId: c.systemID,
+			NginxId:  c.nginxID,
+		},
+		Zconfig: config.Config,
+		Zaux:    config.Aux,
+		DirectoryMap: &proto.DirectoryMap{
+			Directories: config.Directories,
+		},
 	}
+
+	payload, err := json.Marshal(cfg)
+	if err != nil {
+		c.logger.Error(err, "failed to send config")
+		return err
+	}
+
+	metadata := &proto.Metadata{
+		Timestamp: types.TimestampNow(),
+		MessageId: request.GetMeta().GetMessageId(),
+	}
+
+	payloadChecksum := checksum.Checksum(payload)
+	chunks := checksum.Chunk(payload, 4*1024)
+
+	err = downloadServer.Send(&proto.DataChunk{
+		Chunk: &proto.DataChunk_Header{
+			Header: &proto.ChunkedResourceHeader{
+				Meta:      metadata,
+				Chunks:    int32(len(chunks)),
+				Checksum:  payloadChecksum,
+				ChunkSize: 4 * 1024,
+			},
+		},
+	})
+
+	if err != nil {
+		c.logger.Error(err, "failed to send config")
+		return err
+	}
+
+	for id, chunk := range chunks {
+		c.logger.Info("Sending data chunk", "chunk ID", id)
+		err = downloadServer.Send(&proto.DataChunk{
+			Chunk: &proto.DataChunk_Data{
+				Data: &proto.ChunkedResourceChunk{
+					ChunkId: int32(id),
+					Data:    chunk,
+					Meta:    metadata,
+				},
+			},
+		})
+
+		if err != nil {
+			c.logger.Error(err, "failed to send chunk")
+			return err
+		}
+	}
+
+	c.logger.Info("Download finished")
+
+	return nil
 }
 
 // run is a blocking method that kicks off the connection's receive loop and the CommandExchanger's Run loop.
 // run will return when the context is canceled or if either loop returns an error.
 func (c *connection) run(parent context.Context) error {
```

Comment:
I don't know. They aren't used outside of the package.

Reply:
I see you point.
Let me better explain my concern.

It is hard to tell  that `sendConfig()` and `receiveFromUploadServer()` are called (1) externally and also (2) in a separate goroutine, which in turns makes it more difficult to understand the connection. especially, from the point of concurrency.

Additionally, there is a bit inconsistency (prob because the PR is draft)

Commander uses:
- ID()
- State()
- run()
- sendConfig()
- receiveFromUploadServer()

Reply:
Makes sense. What about an unexported interface that contains the methods the Commander uses?

Reply:
Would it help if `connection` was in a separate package?  I kept `connection` and `commander` together because there's a strong dependency between them. But if I introduce an interface, I could easily separate them.

Reply:
> Makes sense. What about an unexported interface that contains the methods the Commander uses?

this makes sense to me

> Would it help if connection was in a separate package? I kept connection and commander together because there's a strong dependency between them. But if I introduce an interface, I could easily separate them.

if there is  a strong dependency, not sure if separating them into different packages is reasonable